liquibase-core/src/main/java/liquibase/snapshot/jvm/ColumnSnapshotGenerator.java

```diff
@@ -184,6 +186,98 @@ protected void setAutoIncrementDetails(Column column, Database database, Databas
         }
     }
 
+    /** Discriminate (driver doesn't) between columns with backing sequence and columns with SERIAL data type */
```

Comment:
Why do we have to discriminate between columns created as serial and ones that have a backing sequence? The serial type is just an alias for creating the backing index and a big part of what we try to do in the generateChangeLog process is try to capture a semantically correct state of the database and we should just be able to treat a column as autoIncrement/serial whether it was created that way or switched to be that way later. 

Is the main problem that the generated changesets have both a create sequence and then an attempt to create the column as serial that causes the sequences to be mirrored? If so, the better fix is maybe in the MissingSequenceChagneGenerator to check if the column is auto-increment and then just not create the sequence?

Reply:
> Is the main problem that the generated changesets have both a create sequence and then an attempt to create the column as serial that causes the sequences to be mirrored?

yes, in simple terms that's the case.
> If so, the better fix is maybe in the MissingSequenceChagneGenerator to check if the column is auto-increment and then just not create the sequence?

I did contemplated this move. as well as reverse case -- throw away all SERIALs and always create sequence (with CREATE IF NOT EXISTS appended) -- and that's has deeper cause that even Postgres doesn't recognize SERIAL data type as first class citizen: if you do plain text backup for schema or use describe table command in PSQL -- you'll get int/bigint column with applied sequence for default value. the only difference between two, at the end of the day, is that one sequence was create by system and another -- by user. and my fear was that if you'll do backup/restore -- you'll lose even that information (but looks like binary backup with data works as expected; at least I failed to find any inconsistencies)
why I opted out? well, there is no single major reason, but in short it just looks wrong.
\- if customer would like to use SERIALs then they are free to do this (and that's what I would do in their place)
\- but if they wan't to use sequences... I don't see reason why we shouldn't allow them. and on top of that we do -- you can create, alter, rename and drop one; the mess begins when you try to use this sequence in table, but Datical ignores you. then why would you create one? why we snapshot sequences in the first place? (to later escape them if they got used)
\- code, in its messy way, is just trying to sort all this out. if our assumptions are wrong -- we'll use SERIAL type eventually. at least this was the intention.

on the other hand, I'm not happy with changes in this PR myself.
mostly because, if we try to fix every little thing with alike approach -- Liquibas will quickly become undigestable.
but in this case most of the blame is behind Postgres: they don't give you information you need (like sequences don't know about columns that they belong to; columns know that they are of autoincrement type but the fact that they are behind sequence -- default value expression should be parsed; as for trying to retrieve real column type -- I spend half a day and just gave up in frustration); and some information is just misleading.

Reply:
as a late afterthought, I think there is one viable alternative approach: we can retrieve all SERIAL sequences and SERIAL sequence knows about column its applied on (actually we need only SERIALized column list)
and if column that we processing not in the list -- then we assume that sequence for this column exists.
disadvantages:
\- with current implementation we rely on existing code (we just reuse cached sequence lookup); in this case we must write additional functionality
\- above logic undercover relies on `OWNED BY` clause. if one would alter system generated sequence - we end up trying to apply default value that references non-existing entity (not a highly probable scenario, sure, but...)

Reply:
It helps to remember that the snapshot and the diffChangeLog operations are completely independent. The job of snapshot is to just capture all the information in the database into an internal data model we can use. The diffChangeLog operation uses that snapshot information to try to figure out the best way to create a changelog that would get a different database to match that snapshot state using our standard change tags. That changelog generation process can do whatever it wants to make the "best" changelog and that may end up being different than what was originally used to create the objects and if our change model cannot handle all the options on the original object some may be lost. 

In this case, I'm thinking that during the snapshot process, since serial columns don't really exist in postgres we shouldn't be trying to reverse-engineer them in there. We should just be capturing the columns, default values, and sequences like we normally would. Then, when it comes time to create the changeSets from that snapshot, the diffChangeLog logic needs to be smart enough to recognize that a sequence is there as backing an auto-increment column and so not manually create that sequence and instead mark the column as autoincrement. That way when you run the changelog it will correctly create the sequence as owned by the column. 

There are some edge cases there like if they manually created a backing sequence with an odd name, but I'm not sure how often if ever that actually happens for us to make sure the sequence names match the serial-generated names or not.

Reply:
just to make sure we on the same page:

\- can you confirm what output we expect for next cases:
```
-- case1
-- script
create table serial_table(id SERIAL);
-- should be converted into
create table serial_table(id SERIAL);
```
```
-- case2
-- script
create sequence some_sequence;
create table some_table(id int default nextval('some_sequence'));
-- should be converted into
create table some_table(id SERIAL);
```
```
-- case3
-- script1 (first run of deploy packager)
create sequence other_sequence;
-- script2 (second run of deploy packager)
create table other_table(id int default nextval('other_sequence'));
-- should be converted into
??
```

\- no changes (or minimum) at snapshot level are expected. 
one should move decision to `Missing/Changed<Entity>ChangeGenerator` level.
here, I'm not sure where I should get data to make this happen.
is it possible to do with existing interfaces or proposed changes in DAT-1926 are required (add ability to access `snapshot` at fixMissing() level)?

----

we did hit some sequence name conflicts.
they are not as rare as maybe expected.
the reason is that Postrges generates value is very human-readeable in itself. second case -- people very often apply some sort of plain database backups, so this name pattern very often appears in our tests for example.
behavior depends on order. if sequence was created first, then SERIAL will create sequence with the same name but append **1** prefix.
in reverse order case (and we hit this one after adding Postrgres dependency sorting -- sequence is not used so it goes in tail of missing object list) -- statement will fail (for this reason Roman added 'CREATE IF NOT EXISTS...' for Postgres generated statements)

Reply:
Yes, that was the cases I was thinking of, other than thinking that the OWNED BY is an important part of determining if it's a SERIAL column as well, not just the default value pattern. 

But, you are right that it is easy for the sequence names to drift and you do run into problems with case 3 when you first create the sequence and later make an serial column that uses it since the generated auto-increment column will use a different sequence. Another time the sequence names may drift is when a table is initially created and then renamed. The sequence will continue to keep the old name.

I'm torn on the best way to handle it. From a diffChangeLog design standpoint, I'd say we should be modeling the fact that the column is auto-increment because that is what allows the generated changelog to be most semantically correct and cross-database. You may get differences in sequence names, but that is fine just like you get differences in generated foreign key constraint names. There are also some edge cases where you can end up with extra sequences in your system (like your case 3) but the normal liquibase response is to say "fix up your changelog by manually removing that extra sequence". 

However, because Datical hits those edge cases more often and doesn't like people working with the changelog file, that makes that "follow the design" approach more difficult. To correctly handle the use cases, we have to be more postgresql-specific and just create the sequences, make the columns, then mark the sequences as owned by that column all the time. So whether they make the column as serial or set it up manually, you always get a <createSequence> changeSet, then the createTable/addColumn changeSet with a defaultValue defined, then an <alterSequence associatedWith=X> changeSet. That is a longer series of changes and very much postgresql specific, but will handle all the edge cases.